Support the Ltac debugger in CoqIDE

jfehrle opened the pull request:
~~This is a prototype for evaluation (first by @gares) of the XML protocol changes for supporting the debugger in Coq.  It works but there are still lots of GUI details to address.~~

**UI**: The Messages panel is used as a terminal, showing the debugger output and letting the user enter debugger commands at the bottom of the panel, like this:

![image](https://user-images.githubusercontent.com/1253341/105624201-ae151800-5dd4-11eb-9513-c66c4a93c1b8.png)


**Current implementation:** coqidetop is single threaded.  It has a message read-eval-print loop, used to accept the usual Coq tactics and commands.  In coqtop, the debugger has a separate read-eval-print loop that accepts a handful of debugger-specific commands.  You must exit the debugger to enter more Coq tactics and commands.  An unusual feature of the Coq debugger is that control of the debugger and execution are in the same thread--it is more common to have these be separate threads or processes.  That makes it very synchronous in behavior.

**Protocol Requirements:** A debugging protocol should support structured data.  For example, a request to fetch the call stack with filename and location data should be returned as structured data--such as in XML or JSON or whatever LSP/VsCoq uses.  We'll want to allow different GUIs to use different packaging, but each GUI should only have to use one packaging mechanism.  For CoqIDE we'll stick with XML.

**The approach taken in this PR is:**
- Create a single-message read-eval-print routine that is called by both the coqidetop loop and by the debugger when it needs a command.  Given the synchronous and modal (regular messages vs debug) nature of the communication, it's easy to ensure that the messages are read at the appropriate point in the code (top loop vs debugger).
- Modify some of the CoqIDE logic to permit sending and receiving debug requests while a proof-modifying operation in still in progress.  In effect, debug requests are nested inside of proof requests, e.g. this sequence of operations:
1. "Forward command" request -> server
2. Client <- debugger prompt
2. Debugger request -> server
3. Client <- debug reply and output (repeat 2-4 as many times as needed)
4. Client <- forward command reply
- Under the current protocol, coqidetop can send "feedback" (really just output) messages at any time.  We can use this without change for debugger output.
- Messages are sent through a single connection.  A second connection isn't needed to handle asynchrony.
- Multiple protocols for different GUIs can be supported by making the read-eval-print loop pluggable.

The protocol code is in the final commit.  Let me know your thoughts and questions.

cc: @gares

gares commented:
The first two improvements I would suggest are:
- get a pointer from @coq/engine-maintainers about where to put the debug channel in the monad, avoiding the long spaghetti. I'm not against a global setting, but the toplevel should have an API to install a debug channel inside the ltac interpreter.
- use a 4th tab, say Ltac Debugger, to avoid tampering with the message tab code (eg the clear/don't clear things). Feedback messages do carry a route id (an integer, always 0): pick 1 and route all these feedbacks to the debug tab, which shall have its own clear/don't clear policy

jfehrle commented:
I had trouble figuring out how to get anything working with the monads, so I'm certainly interested in suggestions on that part of the code.  

It would be helpful to have a "Monads for Dummies" developer writeup explaining how to convert non-monadic code to monadic form, how to delay or not delay evaluation of certain parts, etc.  I expect the rules are fairly simple but it shouldn't be necessary for every developer to reverse engineer the code to understand it.  I could create such a writeup if someone knowledgeable helps me get it right and complete.

For the messages tab code, my intention is to retain the current behavior for non-debug operations (i.e. always clear the panel) but to not clear it for debugger operations.  I'm not keen on adding more panels--it would mean the user has to look in 2 different places to see the result of their last action.  Also not so great for the layout and it means.  I'd like to continue with that original plan for now and have people try it first before changing course.

gares commented:
What I was suggesting is exactly to not sprinkle if-debug conditionals here and there in order to make an already working widget serve a different purpose. It is the same gut feeling I have about using the same channel and the same protocol for two purposes. It can be made to work, but I think it results in code that is harder to understand.
We already have a few instance of this "anti pattern" in the code base, see for example the beautifier: one cannot even answer the question "where is it implemented?" in a satisfactory way, because the answer is "a bit everywhere".

Given the debugger is "modal" I don't see what is the problem in opening a new window (or selecting a new detachable tab) when the debugging session starts. And I don't see why one would need to look at the Messages window during a debugging session if debug messages are routed elsewhere.

SkySkimmer commented on engine/logic_monad.ml:
```diff
@@ -96,6 +99,8 @@ struct
     let (e, info) = Exninfo.capture e in
     raise (e,info) ()
 
+  let get_debug_cmd = fun () -> (); fun () -> Pervasives.(!)forward_read_debug_cmd ()
```
```suggestion
  let get_debug_cmd = fun () -> (); fun () -> forward_read_debug_cmd.contents ()
```
avoids the deprecated Pervasives.

Alternatively put the ref and getter before `!` is shadowed in the file.

ppedrot replied:
You shouldn't even put this in this sealed module. It should only be defined somewhere in the Ltac plugin.

jfehrle replied:
I've tried several variations, but I haven't been able to get the routine to compile if I move it into `tactic_debub.ml`:

I get this error for the following code:

```
File "plugins/ltac/tactic_debug.ml", lines 175-179, characters 4-7:
175 | ....fun () ->
176 |       (match Logic_monad.forward_read_debug_cmd.contents with
177 |       | Some f -> f
178 |       | None -> failwith "forward_read_debug_cmd")
179 |      ()
Error: This expression should not be a function, the expected type is
       string Logic_monad.NonLogical.t


module S =
struct
  open Logic_monad.NonLogical

  let get_debug_cmd : string t =
    fun () ->
      (match Logic_monad.forward_read_debug_cmd.contents with
      | Some f -> f
      | None -> failwith "forward_read_debug_cmd")
     ()

end

    (if is_coqIDE () then
      S.get_debug_cmd
    else
      Proofview.NonLogical.catch Proofview.NonLogical.read_line
      :
```

If I remove `: string t`, I get

```
File "plugins/ltac/tactic_debug.ml", line 193, characters 6-21:
193 |       S.get_debug_cmd
            ^^^^^^^^^^^^^^^
Error: This expression has type unit -> string
       but an expression was expected of type 'a Proofview.NonLogical.t
```

As for `forward_read_debug_cmd`, IIUC if it is defined in `tactic_debug.ml` in the plugin then it can't be accessed from `idetop.ml`, in which case it some other file that's not closely related.

ppedrot replied:
Use the `make` function, or wrap stuff with the monadic return. In any case, this should *not* belong to the low level module defining monadic primitives since it is 1. derivable from the other constructs and 2. Ltac specific. I don't know if there is a fully satisfying solution that does not involve linking idetop to the Ltac plugin, but at least put it a the highest possible point in the dependency chain, not the lowest one.

jfehrle replied:
Thanks.  I was modeling my routine on `read_line`, which doesn't use `make`.

> this should *not* belong to the low level module defining monadic primitives ...

As I noted though perhaps not so clearly in logic_monad.ml:

![image](https://user-images.githubusercontent.com/1253341/106042604-1f352380-6092-11eb-85c8-5a6290d09edf.png)

> I don't know if there is a fully satisfying solution that does not involve linking idetop to the Ltac plugin, but at least put it a the highest possible point in the dependency chain, not the lowest one.

It has to be below `tactic_debug`, which is likely to be fairly low level.  For the moment I put it in `Dyn`, which makes some sense since the plugins are dynamically loaded code.

SkySkimmer commented:
>get a pointer from @coq/engine-maintainers about where to put the debug channel in the monad, avoiding the long spaghetti. I'm not against a global setting, but the toplevel should have an API to install a debug channel inside the ltac interpreter.

What do you mean? The PR touches mostly ide/ (which I didn't look at and doesn't care about the monad AFAIK), I don't see the long pasta.

gares commented:
> What do you mean?

You found it, it is the first hunk: that ref is set in coqidetop. From the toplevel to engine qualifies as a long spaghetti to me ;-)
I don't know if one can completely avoid that, but at least there should be a documented API for setting it.

jfehrle commented:
> What I was suggesting is exactly to not sprinkle if-debug conditionals here and there ...

You're getting well ahead of me.  I said the GUI was not complete.  There are as yet no such conditionals in the message panel code.  Perhaps you would wait for me to propose code I consider fairly complete before complaining about it?

From your [earlier comment](https://github.com/coq/ceps/pull/53#issuecomment-759971172) I understood that you were OK using the message panel for the debugger.  It's fine to change your mind but I'm not yet done looking at that idea.

ejgallego commented on lib/feedback.mli:
```diff
@@ -15,6 +15,7 @@ type level =
   | Debug
   | Info
   | Notice
+  | Prompt
```
Umm, is this supposed to stay? I'm not sure this is the right place as feedback is usually attached to sentences and plays a quite different role.

jfehrle replied:
"Notice" appends a newline to every message.  It seems lame to put a newline after the debugger prompt "TcDebug > ".  I don't like the idea of scanning the output in CoqIDE and suppressing the newline when it matches the debug prompt.  How would you handle it?

BTW, if we create a "Debug" panel separate from the Message panel, then IIUC from @gares, I would need a way to pass a "route id" so the output ends up there.  That might mean adding a "Debugger" category or, maybe better, a labelled "route" parameter on `Feeback.msg_notice` et al.

ejgallego replied:
Feedback shouldn't be used for the debug channel, I think, as it corresponds more to the notion of diagnostic found in common UI protocols.

jfehrle commented:
> use a 4th tab, say Ltac Debugger, to avoid tampering with the message tab code (eg the clear/don't clear things). Feedback messages do carry a route id (an integer, always 0): pick 1 and route all these feedbacks to the debug tab, which shall have its own clear/don't clear policy

I updated the code.  The clearing was not necessary so I removed it, so there's no change in behavior now unless you enter the debugger.

Currently `idtac` output goes to the Messages panel.  Putting the debug output in a separate Debugger panel seems like it would be considerably more awkward for users--they'd likely have to ping-pong between the two panels.

The changes to the Messages panel code are, I think, localized and easy to understand.  Having a separate Debugger panel would mean subclassing the Messages panel code to reuse that code rather than just duplicating it with a naive cut-and-paste.  Probably we'd have to refactor that file so it hews more closely to object-oriented style.  The overall code won't be any smaller; it may be a larger.

@gares, please take a look at the code and let me know what you think.  Maybe try running it, too.

BTW, I fixed a bug in the detach/attach code for Messages panels, a one-line change in `session.ml`, in which the panel wasn't resized to fit into the main window, hiding the scrollbar and the last part of the contents.

ppedrot commented on clib/dyn.mli:
```diff
@@ -93,3 +93,6 @@ sig
 module Make () : S
+
+(* forwarding variables used by plugins *)
+val forward_read_debug_cmd : (unit -> string) option ref
```
This has nothing to do there. Dyn is a module to define dynamic, extensible types. This hook should live as high as possible in the dependency tree. Somewhere in `engine/` might be reasonable.

jfehrle replied:
I moved it to `engine/DebugHook.ml`.

ejgallego replied:
Still seems in the wrong place, who is the user of this forward declaration?

Also, you didn't delete the thing in `dyn.ml`.

jfehrle replied:
I fixed Dyn.  The variable is not in `tactic_debug.ml` because AFAIK `idetop.ml` can't statically link to files that under the `plugins` directory.  If there's a way to do that (and it's a good idea), let me know.

ejgallego replied:
Indeed the fact that this is a plugin makes it quite tricky, @ppedrot , what do you think? Should we add a registration mechanism ? This is already possible to do properly [for example it is what serapi does to load the seralization for plugins, but a bit cumbersome, so I dunno, I think the right place to put this would be `mltop` for now and don't indeed complicate so much with the registration.

gares commented:
> Currently `idtac` output goes to the Messages panel, which some may use for debugging. Putting the debug output in a separate Debugger panel seems like it would be considerably more awkward for users--they'd likely have to ping-pong between the two panels.

Why would you look at the output panel if you are running a debugger? I guess you can print which value an Ltac variable holds without inserting prints (idtac) in your code.

It makes sense for idtac to print to the message windows, but I don't see why this is an argument in favor of mixing the output of your prints with your debug session. On the contrary, if you happen to have any other verbose prints on, eg type class resolution, it is going to flood your debug terminal.

Anyway, if the coqide maintainers are happy with this choice, I'm fine as well.

jfehrle commented:
> It makes sense for idtac to print to the message windows, but I don't see why this is an argument in favor of mixing the output of your prints with your debug session. On the contrary, if you happen to have any other verbose prints on, eg type class resolution, it is going to flood your debug terminal.

My thought is that print statements and debug output are complementary ways of indicating the sequence of operations executed.  If you separate them then it's no longer easy to relate `idtac` output to debugger output.

It's unclear to me how often a user will need to look at any debugger output before that related to the current breakpoint (i.e. the bottom of the panel).  I tend to think not that often.

Keep in mind that this is the first in a series of incremental PRs.  It's not clear to me that we'll want to keep a Debug panel around for the long term.

The current coqtop debugging techniques can generate a huge amount of output. Supporting breakpoints in the debugger will make it possible to skip over large chunks of code, meaning far less single stepping and much less output--and making it practical to debug more complex code.  Instead of printing the code about to be executed, it would be better to highlight that code in the script panel.  The goal could be shown in the goal panel rather than in the Debug or Messages panels.  If we later choose to make debugger commands available only through menu items and function keys, then there wouldn't be any point in having a Debug panel.

FWIW, for this initial PR, I'm trying to get to parity with debugging in coqtop, which has everything come out in a single stream.  For the short term.  I expect debugging in coqtop will used be much less (not at all?) as CoqIDE and others begin to support the debugger.

I'd like to get more of the basics in place and have a full discussion of the use cases for it (at a later date) before considering adding a separate Debug panel.

ejgallego reviewed (dismissed):
Most things look good to me, however the use of Feedback plus ad-hoc hook for the communication with the debugger must be replaced by a dedicated API which can be used by IDEs.

Fortunately this is not hard to do so IMHO it is reasonable to implement it before the merge.

jfehrle commented:
> Feedback shouldn't be used for the debug channel, I think, as it corresponds more to the notion of diagnostic found in common UI protocols.

Your suggestion is to create separate API functions and XML messages to send debug output to CoqIDE?  That would just make the code larger for no benefit.  If VsCoq protocol, for example, requires separating debug output, that separation is best done inside `print_debug` and `print_prompt`.  I don't see an advantage to adding this functionality now over waiting until VsCoq support is being integrated--when we can readily verify that the additional code does exactly what VsCoq needs.  

> ad-hoc hook for the communication with the debugger must be replaced by a dedicated API which can be used by IDEs.

I'm happy to add a setter function instead of using `DebugHook.forward_read_debug_cmd := Some read_debug_cmd;` if that's what you mean.  If you mean more than that, would you elaborate a bit?  Otherwise, it's better to do this refactoring when VsCoq is being integrated and we can readily verify that the API does what's needed.  (It seems quite premature to restructure `coqide.ml`, if that's what you're suggesting.)

Let me know your thoughts.

ejgallego commented:
> Your suggestion is to create separate API functions and XML messages to send debug output to CoqIDE? That would just make the code larger for no benefit.

I am talking about the ML-level API, how to encode for the server-client communication is larguely independent.

An indeed there is a benefit, in terms of keeping the API sane and not mixing two totally different notions which would confuse other clients.

> If you mean more than that, would you elaborate a bit?

I mean indeed providing a well-defined API , likely `Tactic_debug.register_debugger` , which allows clients to install the debugger hooks. BTW I didn't look at this in detail but it seems to me that the hook you currently register is in the wrong place?

jfehrle commented:
Updated.  I renamed `msg_prompt` to `msg_debugger_prompt`, etc.  I think that name is clear enough to  discourage inappropriate use.  Do you think that's sufficient?  Do more?  Something different?

`msg_debug` is used all over the code, it's not at all specific to the debugger.  Some of the debugger output is `DEBUG`, some is `NOTICE` and, of course, one is `DEBUGGER_PROMPT`.  This can easily be tweaked so all debugger output is identified with debugger-specific levels.

I added debugger hook routines.

ejgallego commented:
> Do more? Something different?

I think indeed we need to further tweak the registration so:
- msg_prompt goes away in favor of a channel / function added in the hook
- debugger commands [and answers] are reified into an ADT that reflects what the protocol is
- coqidetop and coqtop can install then different debugger handlers

jfehrle commented:
Updated 

> I think indeed we need to further tweak the registration

This is considerably more than a tweak.  It took me several hours.  The routine to parse a debugger string into an ADT can't be in `tactic_debug.ml` because it wouldn't be accessible outside of the Ltac plugin.  So it's in `debughook.ml`.

I don't understand all the details of the monads.  I had considerable trouble getting the parsing code to compile in `debughook.ml`, so I made it less monadic.  If it's not right, which wouldn't surprise me, I'll need very explicit instructions on what changes to make.  Or perhaps you would tweak I did?

```
let string_get s i =
  try Proofview.NonLogical.return (String.get s i)
  with e ->
    let e = Exninfo.capture e in
    Proofview.NonLogical.raise e

```

ejgallego commented:
The whole setup is tricky as indeed the API as of today needs the rework I explained in previous messages, in particular:

- place the registration API in mltop for now,
- the hacks such as "are we in coqide" should just go away, in favor of each toplevel registering the handler
- the handler should be of the following form
  + first, types for messages should be provided, something such as `module Debugger = ... type cmd = ... type answer = ....`, these should be algebraic data types
  + second, the actual hook should reflect the semantics you intend, for example a fully synchronous setup would provide `send : cmd -> answers`
  + then each frontend for the debugger can handle the interaction in the way the need, it is going to be very different depending on the nature of the IDE

jfehrle commented:
> the API as of today needs the rework I explained in previous messages

I've addressed those points already, but you keep changing your mind.

> the hacks such as "are we in coqide" should just go away, in favor of each toplevel registering the handler

This was already removed.  Please read the code.

> the handler should be of the following form

These changes are not needed for this PR and should not be a reason to delay this PR.  Perhaps they will be useful when the next debugger alternative is integrated, which probably is a long while from now.  It's better to wait until we understand what, say, VsCoq needs when we're integrating it rather than guessing what might be useful and probably getting it wrong.

Your details don't make much sense.  They seem pretty speculative:

>  + first, types for messages should be provided, something such as `module Debugger = ... type cmd = ... type answer = ....`, these should be algebraic data types

What's the benefit?

>  + second, the actual hook should reflect the semantics you intend, for example a fully synchronous setup would provide `send : cmd -> answers`

Synchronous??  Suggesting that there is an alternative asynchronous setup?? What are you thinking?  There is no return value for any of the current operations.  Once you enter the `prompt` routine, it processes multiple commands until the entire Ltac tactic has been executed.

>  + then each frontend for the debugger can handle the interaction in the way the need, it is going to be very different depending on the nature of the IDE

This should be addressed as each frontend is integrated, when we have concrete knowledge of the requirements rather than guessing possible abstractions so far in advance.  Can you give concrete examples of those big differences?

ejgallego commented:
> I've addressed those points already, but you keep changing your mind.

Can you point out where and when I have changed my mind?

ejgallego commented on engine/debugHook.mli:
```diff
@@ -0,0 +1,32 @@
+  print_notice : Pp.t -> unit;          (* print a notice *)
+  print_debug  : Pp.t -> unit;          (* print a debug *)
+  print_prompt : Pp.t -> unit           (* print the debugger prompt *)
+}
```
Maybe we could structure the responses here a bit more? I'd suggest you add a type `debugger_response` and let the backends print it; I'm surprised that the debugger would for example output with different "levels".

jfehrle replied:
In coqtop, "debug" gives you this lovely purple tag:

![image](https://user-images.githubusercontent.com/1253341/109760789-27f2ab00-7ba4-11eb-9310-59b76d128fed.png)

ejgallego commented on ide/coqide/idetop.ml:
```diff
@@ -73,8 +73,8 @@ let ide_cmd_checks ~last_valid { CAst.loc; v } =
-  if is_debug v.expr then
-    user_error "Debug mode not available in the IDE"
+  if false then
+    user_error "Dummy"
```
This needs to go.

ejgallego commented on lib/feedback.ml:
```diff
@@ -14,6 +14,7 @@ type level =
   | Debug
   | Info
   | Notice
+  | Prompt (* like Notice but without adding a new line at the end *)
```
Needs to go.

ejgallego commented on plugins/ltac/tactic_debug.ml:
```diff
@@ -12,6 +12,7 @@ open Util
 open Names
 open Pp
 open Tacexpr
+open DebugHook
```
DebugInterface maybe is a better name?

jfehrle replied:
Maybe, but does it really matter?  If you really, really want it...

ejgallego commented on toplevel/coqtop.ml:
```diff
@@ -167,6 +167,11 @@ let coqtop_init ({ run_mode; color_mode }, async_opts) injections ~opts =
+    { read_cmd     = (fun () -> DebugHook.parse_cmd (read_line ()));
+      print_notice = Feedback.msg_notice;
+      print_debug  = Feedback.msg_debug;
+      print_prompt = Feedback.msg_prompt });
```
The last 3 should be replaced by the toplevel printer for debug answers.

ejgallego commented on vernac/topfmt.ml:
```diff
@@ -168,6 +170,7 @@ let default_tag_map () = let open Terminal in [
     "message.error"    , make ~bold:true ~fg_color:`WHITE ~bg_color:`RED ()
   ; "message.warning"  , make ~bold:true ~fg_color:`WHITE ~bg_color:`YELLOW ()
   ; "message.debug"    , make ~bold:true ~fg_color:`WHITE ~bg_color:`MAGENTA ()
+  ; "message.prompt"   , make ~fg_color:`GREEN ()
```
all the changes in this file can go

jfehrle replied:
I had changed the debugger prompt to be green in both CoqIDE and coqtop to make it easier for users to notice the prompt.  I hope we can keep that detail.

jfehrle commented:
I think we're a little closer to agreement, but I have to disagree on some points:

> * the changes to feedback need to be removed in favor of the communication channel. Prompt just doesn't fit into the feedback paradigm, for example breaks the correspondence of feedback with LSP error diagnostics

`idtac` messages (printed with `Feedback.notice`) and debugger output are complementary ways for users to observe the sequence of operations.  We shouldn't separate them in separate panels in the GUI (or send them over separate channels).  That would just make it harder for the user to understand.  The current print_* callbacks aren't helpful for this PR or CoqIDE.  They are, however, sufficient for language servers to create separate channels if they wish.  Doing so is beyond the scope of this PR.

The LSPs can avoid ever generating a "Prompt" message simply by mapping `print_prompt` to the same thing it sets `print_notice` to.  Or something else.

> after doing a pass on tactic_debug IMHO we should introduce a type of answer as to ensure consistency.

This refactoring is not needed for CoqIDE, so it should not be done in this PR.  There is no time savings from doing it now.  Note that debugger commands can generate more than one response message for printing, for example for "r 5" and probably even for single stepping.  We use standard print routines that generate Pp.t data structures.  Changing Pp.t's to a different format is definitely out of scope.

> Jim, one interrogation sign is enough. Synchronous vs asynchronous here are in the context of typical RPC, in one case you wait for the call to return, in other case you don't and handle the answers with a delay.

I'm familiar with the technology.  `prompt` is by nature synchronous.  If an LSP needs to use asynchronous RPCs, that should be handled in the callbacks such as `read_cmd`, hidden from `tactic_debug.ml`.

> I'd dare to say that having implemented 5 different language servers that have been used to build a few different frontends I got a good idea on what the requirements are.

Then you are the best person to generalize what has been done so far when language server integration begins.  Making me the intermediary is a bad idea.

> Regarding the exceptions in the monad

Let's come back to this later--I need to look at the code some more, I hoped you could point out specific details that need to change.  I'm not expert on monads.

jfehrle commented:
One other thought: keep in mind that this is only the first PR in a series.  The printed debugger output will become less useful over time in CoqIDE.  For example, the next PR highlights the code being executed in the script window, making it pointless for the debugger to print that code in the messages window.  Similarly, instead of printing the goal in the debugger panel, we should probably display it in the goal panel (CoqIDE would request the current goal when it reaches a breakpoint).

gares commented:
@jfehrle you could open your next PRs as drafts, if you did not already.

I believe this PR should be discussed at the CoqCall and that the other @coq/coqide-maintainers should have a look at it before that discussion.

@jfehrle @ejgallego if you think this is a good idea, would you schedule this item in a Coq call that fits your schedule?

jfehrle commented:
> @gares you could open your next PRs as drafts

I only have code for the next PR, the others I mentioned outside this thread are merely planned.  That PR is still missing some functionality and has probably needs obvious cleanup in places.  If you'd like to see it in its current state I will post it, otherwise I would wait until I've done as much as I can.

> I believe this PR should be discussed at the CoqCall

How about next week, March 17?

> @ejgallego all clients need to adapt to version bump of the protocol

What are the other clients and why would this be a significant burden on them?  IIUC adding new messages, which this PR does, requires bumping the version number with our current infrastructure.

ejgallego commented:
As discussed in the call today I've done some work to finish this PR, see https://github.com/ejgallego/coq/tree/debugger_coqide

It is almost ready, but it needs a bit more work to be finished.

jfehrle commented:
>  @jfehrle @ejgallego if you think this is a good idea, would you schedule this item in a Coq call that fits your schedule?

@gares suggested we find a mutually agreeable time that fits our schedule.  You're displaying an extreme lack of respect by having a discussion today in today's call before I could even see @gares request.  Let's discuss this when both of us can participate on March 17.

> As discussed in the call today I've done some work to finish this PR, see

I would not spend too much time on this yet.  It doesn't compile, either.

jfehrle commented:
@ejgallego Would you rebase your changes on top of the latest version here and push your additional changes to this PR?  Then I'll verify the user experience is not changed.  Thanks.

ejgallego commented:
Did the rebase but need to do some testing myself, I am pretty busy with deadlines this week so bear with me.

jfehrle reviewed (commented):
This compiles, however the debugger hangs when I enter a command in the coqtop debugger:

```
Coq < Goal True.
Unnamed_thm < Set Ltac Debug.
Unnamed_thm < Ltac x := idtac "x"; idtac "y".
x is defined

Unnamed_thm < x.
```

Some comments follow.

> I am pretty busy with deadlines this week so bear with me.

Sure, no problem.

jfehrle commented on toplevel/coqtop.ml:
```diff
@@ -133,6 +133,20 @@ let print_style_tags opts =
+    CErrors.user_err Pp.(str "ltac_debug_input: " ++ str error)
+
+let ltac_debug_answer = let open DebugHook.Answer in function
+    | Prompt _ -> ()
+    | Goal _ -> ()
+    | Output _ -> ()
+      (* print_notice = Feedback.msg_notice;
+       * print_debug  = Feedback.msg_debug;
+       * print_prompt = Feedback.msg_prompt }); *)
```
```suggestion
    | Prompt pp -> Feedback.msg_notice pp
    | Goal pp -> Feedback.msg_notice pp
    | Output pp -> Feedback.msg_notice pp
```

This gets the coqtop debugger to work.

I see the build fails with

```
File "vernac/debugHook.mli", line 26, characters 36-44:
Error: Unbound module Result
```

but oddly enough `make` works fine for me locally.

ejgallego replied:
Indeed this wasn't complete, fixed, but used direct terminal access.

The point of this code is actually to access the terminal directly, not use feedback which knows nothing about the output backend, so we can have more control including interactive stuff / readline if desired.

ejgallego replied:
Oh, this is due to OCaml 4.05 not having this module, let me fix.

ejgallego commented on vernac/debugHook.ml:
```diff
@@ -0,0 +1,80 @@
+   with their provided interface. *)
+module Action = struct
+  type t =
+    | Step
+    (** Step one command *)
+    | Skip
+    (** Skip one command *)
```
Fixed, thanks.

jfehrle replied:
I don't see the fix.

ejgallego commented on vernac/debugHook.ml:
```diff
@@ -8,68 +8,73 @@
 
-(* todo: didn't understand the point of "Exninfo.capture e"
-   in the monadic form of the following *)
+module Answer = struct
```
I have been using Command / Answer for a long time, but indeed Response may be better.

ejgallego commented on ide/coqide/idetop.ml:
```diff
@@ -541,12 +544,31 @@ let loop ( { Coqtop.run_mode; color_mode },_) ~opts:_ state =
-        print_debug  = Feedback.msg_debug;
-        print_prompt = Feedback.msg_prompt });
+
+  (* XXX: the global ref has to go *)
```
In the future, sorry I was not clear about that.

My requests have usually to do with coding style and coding rules that directly impact the quality of the OCaml code; industrial OCaml code bases such as the ones of Jane Street implement very strict rules at many levels. We are trying to raise the standard.

Functional Programming in OCaml is very different from other languages, and I can see from your code that many times is not idiomatic.

jfehrle replied:
It depends on how the global variables are used and the context for the functions using them.  In the projects I've worked on in industry, reasonable use of global variables was rarely an issue unless there was a functional reason not to, such as when I was writing kernel networking code (e.g. a TCP/IP driver).  Coq and OCaml rely on them in some places.

I wouldn't adopt anyone's coding rules uncritically.  Jane Street's use of OCaml would seem hard to justify from a business perspective--or at the very least not an obvious win.  I'm aware of a wave of other companies eager to adopt OCaml and Jane Street's standards.  In industry, coding rules are secondary to providing useful functionality, code that works correctly and is reasonably maintainable--they are only one tool among many.

But let's not get diverted by this side discussion.  I'm sure we'll have many future opportunities to debate this and other style points.

ejgallego commented on vernac/topfmt.ml:
```diff
@@ -146,7 +145,6 @@ let gen_logger dbg warn ?pre_hdr level msg = let open Feedback in match level wi
   | Debug   -> msgnl_with !std_ft (make_body dbg  dbg_hdr ?pre_hdr msg)
   | Info    -> msgnl_with !std_ft (make_body dbg info_hdr ?pre_hdr msg)
   | Notice  -> msgnl_with !std_ft (make_body noq info_hdr ?pre_hdr msg)
-  | Prompt  -> msgnl_with ~with_nl:false !std_ft (make_body noq info_hdr ?pre_hdr msg)
```
The newline behavior should be preserved by the updated version.

ejgallego commented on plugins/ltac/tactic_debug.ml:
```diff
@@ -38,12 +37,33 @@ type debug_info =
+  let prompt g = wrap (fun () -> (hook ()).submit_answer (Prompt g))
+  let goal g = wrap (fun () -> (hook ()).submit_answer (Goal g))
+  let output g = wrap (fun () -> (hook ()).submit_answer (Output g))
+  let read = wrap (fun () -> (hook ()).read_cmd ())
```
The code above is not idiomatic, in particular when mixing the monad with the ref.

jfehrle commented:
Is this ready for me to try?  I just noticed your change, GitHub didn't send any notification.

ejgallego commented:
> Is this ready for me to try? I just noticed your change, GitHub didn't send any notification.

Indeed I didn't mention you as I still need to connect the reply message to coqide, hopefully I find some time soon, it's been pretty busy days last week with conferences.

jfehrle commented:
Since you've been quite busy recently, how about if we aim to have an agreed-upon mergeable PR by say, April 13 (4 weeks from the Coq call discussion)?  If we're not able to do that, then I suggest we merge my last version as-is and then you could submit your additions in a separate PR.

I'd like to submit 2 more follow-on PRs in time for 8.14.  It's much better to get out a good chunk of functionality rather than dribble it out over multiple releases.  And I'm hoping this doesn't become a very drawn-out project.  Conceptually the changes are not that complicated, although there are many details to address and get right.

ejgallego commented:
> Since you've been quite busy recently, how about if we aim to have an agreed-upon mergeable PR by say, April 13 (4 weeks from the Coq call discussion)? If we're not able to do that, then I suggest we merge my last version as-is and then you could submit your additions in a separate PR.

Unfortunately this won't work as I consider the original PR not mergeable .

Tho I stole a minute and this PR should be now in testable state, I did some basic tests and indeed things seem to work, but given the depth of the changes we may find some issues for sure.

But if testing works fine for you, I don't see anything blocking so we could go ahead.

> I'd like to submit 2 more follow-on PRs in time for 8.14. It's much better to get out a good chunk of functionality rather than dribble it out over multiple releases.

No need to have that discussion here, let's focus on technical issues please.

> And I'm hoping this doesn't become a very drawn-out project. Conceptually the changes are not that complicated, although there are many details to address and get right.

I have to disagree, working on UI is one of the hardest parts in any programming, more the reason to raise the PR review bar.

ejgallego commented:
Before merge: this needs both squashing and fixing the xml protocol changelog.

jfehrle commented on plugins/ltac/tactic_debug.ml:
```diff
@@ -54,24 +79,24 @@ let db_pr_goal gl =
 let db_pr_goal =
   Proofview.Goal.enter begin fun gl ->
   let pg = db_pr_goal gl in
-  Proofview.tclLIFT (msg_tac_notice (str "Goal:" ++ fnl () ++ pg))
+  Proofview.tclLIFT (Comm.goal pg)
```
I would keep printing "Goal:\n" as a separate output before the Comm.goal for the moment.

ejgallego replied:
Sounds good, will amend.

ejgallego replied:
After a second review I think this should be done client-side.

jfehrle reviewed (commented):
I found a few things that need fixing.  Let me know what you suggest.

(I only had time to test it, didn't get to read through your changes.  Tomorrow.)

ejgallego commented on ide/coqide/wg_MessageView.ml:
```diff
@@ -108,17 +177,20 @@ let message_view () : message_view =
       push#call (level, msg)
 
+    method debug_prompt msg =
+      insert_ltac_debug_prompt msg
```
Eventually the debugger I/O should move to its own window I understand? So adding a flag to mark something as a prompt seems fine to me for, will do. However there is still a conceptual problem in the sense of what it does mean for that list to contain two prompts?

ejgallego replied:
In the sense that a prompt is not only a newline, but also it actives edit mode on the window!

jfehrle replied:
After considering it more, it would be cleaner to fix the extra newline by
- having `Comm.output` and `Comm.goal` append `fnl ()` to the message (done once in `tactic_debug.ml` and
- removing the line in `insert_msg` that appends a newline.
Then a 3-tuple is not needed (which is a bit ugly).

> However there is still a conceptual problem in the sense of what it does mean for that list to contain two prompts? ... In the sense that a prompt is not only a newline, but also it actives edit mode on the window!

Detaching the messages window doesn't call `insert_ltac_debug_prompt`, doesn't change edit mode, so not an issue.

> Eventually the debugger I/O should move to its own window I understand?

Still thinking that through.  FWIW I'd like to give the debugger info more structure, e.g. a panel for the stack (click on a line to jump to the code), a panel for variables and a messages/terminal panel.

ejgallego replied:
I've opted to keep the list of messages for now, but indeed I am pretty sure you will like to a custom UI with stacks, etc... I think the protocol / code is in good shape to make this reasonably easy.

ejgallego commented on plugins/ltac/tactic_debug.ml:
```diff
@@ -169,33 +139,22 @@ let run ini =
-  begin
     let open Proofview.NonLogical in
-    Proofview.NonLogical.print_notice (fnl () ++ str "TcDebug (" ++ int level ++ str ") > ") >>
+    Comm.prompt (str "TcDebug (" ++ int level ++ str ") > ") >>
```
Will fix, thanks.

jfehrle commented on plugins/ltac/tactic_debug.ml:
```diff
@@ -37,8 +37,33 @@ type debug_info =
+(* Communications with the outside world *)
+module Comm = struct
+
+  (* TODO: ideally we would check that the debugger hooks are
+     correctly set, however we don't do this yet as debugger
+     initialiation is unconditionally done for example in coqc, which
+     doesn't support the debugger interface. Improving this would
+     require some tweaks in tacinterp which are out of scope for the
+     current refactoring. *)
```
There are any number of ways to break the code.  Only a small number are both easy to check and worth checking, hence the desire to prove code is correct.  If the hooks are not set, the error is really obvious.  Not sure this is worth trying to check.

JasonGross replied:
Note, btw, the it's important to some of my use cases that `coqc` supports `Set Ltac Debug. Set Ltac Batch Debug.`

ejgallego replied:
Umm, can you explain a bit more? We could have `coqc` interact with the terminal but it seems weird for a batch compiler to support interactive debugging; I'm not sure I've heard of any other compiler doing so.

jfehrle replied:
`Set Ltac Batch Debug.` runs the debugger as if the user entered newline at every prompt, so you get an exhaustive transcript of Ltac tactics. I think this compensates for the weakness of the current debugger.

jfehrle commented on plugins/ltac/tactic_debug.ml:
```diff
@@ -37,8 +37,33 @@ type debug_info =
+    | Some _ -> ()
+    | None -> ()
+      (* CErrors.user_err
+       *   (Pp.str "The LTAC debugger is not supported by your user interface") *)
```
```suggestion
       *   (Pp.str "Your user interface does not support the Ltac debugger.") *)
```

jfehrle commented on vernac/debugHook.ml:
```diff
@@ -0,0 +1,80 @@
+    | RunBreakpoint of string
+    | Failed
+
+  (* XXX: These should be removed and std functions used *)
```
What else needs to be done here?

jfehrle reviewed (commented):
Mostly small details

ejgallego commented on ide/coqide/coq.ml:
```diff
@@ -289,20 +293,24 @@ let handle_feedback feedback_processor xml =
 }
 
-let unsafe_handle_input handle feedback_processor state conds ~read_all =
+let unsafe_handle_input handle (feedback_processor, ltac_debug_processor) state conds ~read_all =
```
These are totally different communication channel with totally different specifications, so they are better separate. It also makes testing modular.

ejgallego commented:
Push addressing almost all the comments, thanks.

Note that current master will exit the debugger also with an interrupt, I think it is best to address this in a different PR if desired.

There is still a large chunk of thinks that could be improved, including many parts of both the CoqIDE and tactic_debug architecture, but I think this is OK for now; in particular we could quite improve the parser / prompt protocol, but IMHO that can be done separately, tho it would require a protocol update.

The current architecture should be modular enough as to allow to move the debugger to its own UI with minimal changes.

jfehrle commented:
The protocol change causes the Feedback-based output for `idtac` to appear in the output in the wrong place (delayed after the prompt message).  I've looked at a while.  My _guess_ is that Feedback goes through `Coqops.process_feedback`, which is invoked at 300msec intervals.  The debug output used to go through there but probably doesn't anymore and is gets processed immediately.  So things are out of order.

Jason's case is broken, trace at the bottom.  I fixed a couple things, in particular that "xx" would still throw you out of the debugger that I will push shortly.

Test code:

```
Ltac x := idtac "A";  idtac "B";
 idtac "C"; idtac "D"; idtac "E".
Set Ltac Debug.
Goal True.
x.
```

**Bad output:**

![image](https://user-images.githubusercontent.com/1253341/114259010-b4e90b00-997f-11eb-8a93-e176bd65d442.png)

**Good output:**

![image](https://user-images.githubusercontent.com/1253341/114259047-f7aae300-997f-11eb-90ba-5991384dd9c3.png)

```
Error: Anomaly "Uncaught exception Option.IsNone."
Please report at http://coq.inria.fr/bugs/.
Raised at Option.get in file "clib/option.ml" (inlined), line 50, characters 9-21
Called from Tactic_debug.Comm.hook in file "plugins/ltac/tactic_debug.ml", line 55, characters 16-50
Called from Tactic_debug.Comm.goal.(fun) in file "plugins/ltac/tactic_debug.ml", line 62, characters 31-40
```

jfehrle commented:
I fixed Jason's issue with `Set Ltac Batch Debug.` in coqc and ensured that the output is in the correct order.  So AFAIK everything works correctly and this is about ready to merge.  Take a look before I squash.

jfehrle commented:
@ejgallego There are only a few minor changes in the 2 commits I added recently.  Would you take a look at them?  Shouldn't take you more than a few minutes to review them before I squash.

Will you have time to do this in the next day or two?  Or I could go ahead and squash and ask @ppedrot to finish off the PR.  Please let me know, thanks.

ejgallego commented:
Will do @jfehrle , thanks. There is also a couple of minor nits I'd like to include as to fix some issues you pointed out, and to ensure that the channel can be decoupled from the protocol if desired as was suggested on the CEP.

But indeed the PR is very ready for PMP for review if he wants I think.

jfehrle commented:
> a couple of minor nits I'd like to include as to fix some issues you pointed out

I will squash now and will plan to keep those fixes (the "todo"s I added) as a separate commit in case that stuff is actually needed for some reason.

jfehrle commented:
@ejgallego Will you be able to make your final changes in the next couple of days?  It's taking much too long to finish this PR.  The "todo" items I identified are not related to my changes and don't have to be done in this PR.  That's why I made them todos.  "decoupling the channel" is not needed now and can done later.  @ppedrot can't do a final review/merge until you're done changing things.  Non-essential refactoring should not block functional improvements.

We can discuss on the Coq call if this hasn't been resolved by then.

ppedrot commented on ide/coqide/coqide.ml:
```diff
@@ -765,6 +765,17 @@ let show_proof_diff where sn =
+
+let send_db_cmd cmd = cb_on_current_term (db_cmd cmd) ()
+
+(* technique recommended in https://ocaml.org/releases/4.11/htmlman/comp.html#s%3Acomp-errors
```
IMO this comment is noise, but it's not utterly necessary to remove it.

jfehrle replied:
It was for the benefit of reviewers who strongly objected to a very similar approach in another PR a long while ago.

ppedrot commented on ide/coqide/protocol/xmlprotocol.ml:
```diff
@@ -12,7 +12,15 @@
-let protocol_version = "20200911"
+(** Changelog
+
+    - 20210409: support for ltac debugger:
```
Shouldn't that go somewhere else, e.g. in the `xml-protocol.md` file?

jfehrle replied:
I added info in `xml-protocol.md` and dropped this comment.

ppedrot commented on ide/coqide/wg_MessageView.ml:
```diff
@@ -38,23 +38,35 @@ class type message_view =
   let view = GSourceView3.source_view
     ~source_buffer:buffer ~packing:scroll#add
-    ~editable:false ~cursor_visible:false ~wrap_mode:`WORD ()
+    ~editable:false ~cursor_visible:false ~wrap_mode:`CHAR ()
```
I definitely don't want this change, it's very annoying for readability to get the output cut at random places.

jfehrle replied:
WORD gives unusual behavior for terminal input, but I will change it as you suggest.  The current inputs are very short, so users are unlikely to run into it.  Here are the before and after when I type "Y":

![image](https://user-images.githubusercontent.com/1253341/116006514-7333a380-a5c0-11eb-8e4c-cf908a45d56b.png)

![image](https://user-images.githubusercontent.com/1253341/116006522-7f1f6580-a5c0-11eb-8ae8-2f6173640fc6.png)

I expect things sent as Pp.t would end up word-wrapped with either setting, but some messages are just strings, e.g. `Print Grammar constr`.

ppedrot replied:
My main argument is that your PR shouldn't affect people not using the debugger. Since the debugger widget is likely to be reimplemented in a near future, there is no point in trying to make it better right now while negatively affecting the basic behaviour of the message window.

ppedrot commented on ide/coqide/wg_MessageView.ml:
```diff
@@ -78,15 +90,82 @@ let message_view () : message_view =
+  in
+  let msgs : message_entry_kind list ref = ref [] in
+  let (return, _) = GtkData.AccelGroup.parse "Return" in
+  let backspace = 65288 (*GtkData.AccelGroup.parse "Backspace"*) in
```
How do we know this is stable across architectures? And more, why isn't this done just like in the two lines around?

jfehrle replied:
Got it.  I had to check the GTK source.  I didn't expect it to be "BackSpace" rather than "Backspace".

ppedrot reviewed (commented):
I am vaguely concerned that changes to message view will need to be completely backtracked upon when we implement a proper widget for the debugger. The current situation clearly belongs to the category of prototypes. I don't think it requires changes for the merge though, since we discussed that already. I have a few question-ish nitpicks but I don't think they are necessary for the merge either, depending on @jfehrle 's answers.

ejgallego reviewed (changes requested):
Too much metadata was lost on the squashing, including valuable commit messages and authorship information. It must be fixed before merge.

Additionally, current commit message doesn't adhere to git commit message guidelines.

SkySkimmer commented:
We have commit message guidelines?

jfehrle commented:
> Too much metadata was lost on the squashing, including valuable commit messages and authorship information. It must be fixed before merge.

If you're referring to seeing your name for some of these changes in `git-blame`, AFAIK I would have to keep your commit exactly as it is to show your name, so this would have to be 2 or maybe 3 commits.  Is that worth the trouble?

jfehrle commented:
Updated.  I addressed @ppedrot's comments.  Waiting to hear from @ejgallego and to do a final squash.

jfehrle commented:
> I am vaguely concerned that changes to message view will need to be completely backtracked upon when we implement a proper widget for the debugger.

Part of the concept of incremental development is that the software should always be in a shippable state.  Sometimes that means a bit of wasted motion in the development process.  It should be easy enough to remove some of the message view changes when they're no longer needed.  That could be in the near future (reviews being the rate-limiting step).

ejgallego commented:
Sorry for my succinct message, I was on a rush this Sunday.

> We have commit message guidelines?

@SkySkimmer I was talking about git general convention, which states commit messages have a general summary on the first line, a space, then the rest of the text.

> If you're referring to seeing your name for some of these changes in git-blame, AFAIK I would have to keep your commit exactly as it is to show your name, so this PR would have to be 2 or maybe 3 commits. Is that worth the trouble?

@jfehrle , I think we can improve the commit message a bit, usually when squashing commits that are done by multiple authors we use the non-standard `Co-authored-by`; I'd propose to keep one commit but with a commit message closer to:
```
[ltac debugger] XML protocol support and preliminary CoqIDE UI.

We reify the input/output setup of the Ltac Debugger into a protocol,
so user interfaces can implement their own communication with the
debugger, by registering with the module `DebugHook`.

We provide a preliminary implementation of this protocol for the XML
protocol, (closes #4647) , and a preliminary interaction setup with
the debugger in CoqIDE using the message panel. This is achieved by
setting the message panel to terminal mode when entering the debugger.

Miscellaneous rework of the debugger towards a more functional style
has also been performed [exceptions -> option types, etc...] but some
more work would still be nice to have, in particular
w.r.t. initialization and debugger state.

Co-authored-by: Emilio Jesús Gallego Arias <e+git@x80.org>
```

There are many things to tweak indeed in the new debugger interface but IMO let's do it after merge; I'll be happy to help with some of more hairy things to be done in `ltac_debug`.

Failure of `coqtail` is worrying tho, I may have introduced this but it is likely some strange typo or mistake; I'm afraid I don't understand what is going on, and that is the only thing blocking the merge.

> I added info in xml-protocol.md and dropped this comment.

That's fine, but note what we say in that document "changes are documented in dev/doc/changes.md"; I think we should at least keep a pointer in the code to the canonical location of the changeset [the .md file seems fine, but at some point we could move it to either odoc or sphinx I guess]

ejgallego commented:
My only pending change for this PR was regarding to the comment about the timer and feedback processing, but it is no way a blocker and indeed it seems to me like moot work if we are gonna move the debugger to its own window, which makes sense.

@jfehrle note that while we have been doing joint CoqIDE / Coq releases so far, I'd like to decouple the two, so as long as we are happy protocol wise for 8.14 , improving CoqIDE's interface can be done easily on its own after release, including having your own advanced debugger branch which can be tried with a click for opam users.

jfehrle commented:
OK, I think I fixed the coqtail failure.  Squashed and updated the commit, too.  So ready to commit.  @ppedrot?

jfehrle commented:
@ejgallego Perhaps you mark this as Approved (if it matters)?

jfehrle commented:
> My only pending change for this PR

BTW, this particular wording suggests you want to add more to this PR.  Wording like "My only follow up to this PR" suggests it would be in a later, which IIUC is what you mean.

> note that while we have been doing joint CoqIDE / Coq releases so far, I'd like to decouple the two

I don't want to consider that for the time being--it would make considerably harder to make progress.  At this point adding new features requires changes in both the server and CoqIDE and additional protocol messages--that's much easier to do if they're in the same repository.  It's at least one and maybe 2 more PRs before it may be ready for general feedback, though I'll probably ask a couple people to try it earlier.

jfehrle commented:
BTW, I think the effort to make detailed commit messages is a bit misplaced.  The messages describe what was done in a particular commit on a particular day, but I think their long-term value is small because they can quickly become obsolete.  And they are not so easy by subject later on.  I think it's better to put most of that energy into creating and updating code comments and internals documentation.

ppedrot commented:
I believe that the three CI failures are unrelated to this PR, and due to the switch to the dune-based build system. Therefore let's merge the PR, it has been waiting enough.

jfehrle commented:
Hooray!  Thanks to all those who reviewed and commented.
